File: LLD/src/main/java/org/example/p1_oops/c1_accessmodifiers/p2/StrangerInDiffPackage.java
package org.example.p1_oops.c1_accessmodifiers.p2;

import org.example.p1_oops.c1_accessmodifiers.p1.Parent;

public class StrangerInDiffPackage {
    public void checkAccess() {
        System.out.println("\n--- Inside StrangerInDiffPackage (Different Package, Non-Child Class) ---");
        Parent p = new Parent();

        System.out.println("publicVar: " + p.publicVar + " (Accessible)");

        // System.out.println("protectedVar: " + p.protectedVar);
        System.out.println("protectedVar: NOT Accessible (Protected is visible in diff package ONLY via inheritance)");

        // System.out.println("defaultVar: " + p.defaultVar);
        System.out.println("defaultVar: NOT Accessible (Default is only visible within the same package)");

        // System.out.println("privateVar: " + p.privateVar);
        System.out.println("privateVar: NOT Accessible (Private is only visible within the same class)");
    }
}

File: LLD NodeJs/p1_oops/c1_accessmodifiers/p2/StrangerInDiffPackage.js
const Parent = require('../p1/Parent');

/**
 * Demonstrates access from an unrelated class in a different module/package
 */
class StrangerInDiffPackage {
    checkAccess() {
        console.log("\n--- Inside StrangerInDiffPackage (Different Package, Non-Child Class) ---");
        const p = new Parent();

        console.log("publicVar: " + p.publicVar + " (Accessible)");

        // In JS, there's no enforcement of protected - it's just convention
        console.log("protectedVar: " + p._protectedVar + " (Accessible - JS convention, NOT enforced)");

        // In JS, there's no package-private concept
        console.log("defaultVar: " + p.defaultVar + " (Accessible - JS has no package-private)");

        // console.log("privateVar: " + p.#privateVar);
        // SyntaxError: Private field '#privateVar' must be declared in an enclosing class
        console.log("privateVar: NOT Accessible (Private fields with # are only visible within the class)");
    }
}

module.exports = StrangerInDiffPackage;
